src/event/metric.rs

```diff
@@ -360,6 +379,112 @@ impl Display for Metric {
     }
 }
 
+impl Object for Metric {
+    fn insert(&mut self, path: &remap::Path, value: remap::Value) -> Result<(), String> {
+        if path.is_root() {
+            return Err("cannot set root path for a metric event".into());
```

Comment:
Given the amount of error-cases we'll need to handle here and below, I'd vote to move these into an enum (in Remap we use `thiserror` but we use `snafu` in Vector itself for that).